subsys/logging/log_backend_rtt.c

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2018 Workaround GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <logging/log_backend_rtt.h>
+#include <logging/log_core.h>
+#include <logging/log_msg.h>
+#include <logging/log_output.h>
+#include <device.h>
+#include <rtt/SEGGER_RTT.h>
+
+static int buffer_out(u8_t *data, size_t length, void *ctx)
+{
+	unsigned int key;
+
+	key = irq_lock();
+	SEGGER_RTT_WriteNoLock(0, data, length);
```

Comment:
That would be too easy :)

- If RTT is in non-blocking mode, the SEGGER_RTT_WriteNoLock() could write only part of the buffer or just drop the data completely (it depends on the RTT configuration and state of the buffer). In your code, this will result in dropped/damaged log messages.
- If RTT is in blocking mode, then SEGGER_RTT_WriteNoLock() could spin (in your code with interrupt disabled) several milliseconds until host reads the channel and free some space in the RTT buffer. Such delay definitely disrupt things like Bluetooth or USB. 

IMHO you should:
- Make sure that RTT is in non-blocking mode (i recommend the SEGGER_RTT_MODE_NO_BLOCK_SKIP).
- Handle situations where data is not accepted/partially written by the SEGGER_RTT_WriteNoLock().

Reply:
@nordic-krch, @pizi-nordic Thanks for your feedbacks. Using `SEGGER_RTT_Write()` would allow to not mask higher priority interrupts such as Bluetooth or USB (to be confirmed that their priority level is higher than the RTT one). I would also not need to change `rtt_console.c` that disables all interrupts (that's where my code was inspired) - even if `rtt_console.c` should probably also use `SEGGER_RTT_Write()`, no?

The current RTT library (ie: `ext/debug/segger/rtt`) does not expose a function to know if RTT is blocking or not. But `LOG` subsystem could in theory support partial block transfer, in practical it does not seem to do it - the returned number of written bytes  is never taken in account (look for `ctx->func(` in https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/logging/log_output.c).

Is it correct the changes I need to do are:
- Fix `LOG`  to support partial write
- Use `SEGGER_RTT_Write()` and return the number of written byte

Reply:
Setting the SEGGER_RTT_MODE_NO_BLOCK_SKIP mode is still needed. You can do that through Kconfig.

Reply:
Does `SEGGER_RTT_MODE_NO_BLOCK_SKIP` in KConfig really do something?

```
$ grep -r SEGGER_RTT_MODE_ .
./ext/debug/segger/Kconfig:	default SEGGER_RTT_MODE_NO_BLOCK_SKIP
./ext/debug/segger/Kconfig:config SEGGER_RTT_MODE_NO_BLOCK_SKIP
./ext/debug/segger/Kconfig:config SEGGER_RTT_MODE_NO_BLOCK_TRIM
./ext/debug/segger/Kconfig:config SEGGER_RTT_MODE_BLOCK_IF_FIFO_FULL
./ext/debug/segger/Kconfig:	default 2 if SEGGER_RTT_MODE_BLOCK_IF_FIFO_FULL
./ext/debug/segger/Kconfig:	default 1 if SEGGER_RTT_MODE_NO_BLOCK_TRIM
./ext/debug/segger/systemview/SEGGER_SYSVIEW.c:  SEGGER_RTT_ConfigUpBuffer(SEGGER_SYSVIEW_RTT_CHANNEL, "SysView", &_UpBuffer[0],   sizeof(_UpBuffer),   SEGGER_RTT_MODE_NO_BLOCK_SKIP);
./ext/debug/segger/systemview/SEGGER_SYSVIEW.c:  _SYSVIEW_Globals.UpChannel = SEGGER_RTT_AllocUpBuffer  ("SysView", &_UpBuffer[0],   sizeof(_UpBuffer),   SEGGER_RTT_MODE_NO_BLOCK_SKIP);
./ext/debug/segger/systemview/SEGGER_SYSVIEW.c:  SEGGER_RTT_ConfigUpBuffer   (SEGGER_SYSVIEW_RTT_CHANNEL, "SysView", &_UpBuffer[0],   sizeof(_UpBuffer),   SEGGER_RTT_MODE_NO_BLOCK_SKIP);
./ext/debug/segger/systemview/SEGGER_SYSVIEW.c:  SEGGER_RTT_ConfigDownBuffer (SEGGER_SYSVIEW_RTT_CHANNEL, "SysView", &_DownBuffer[0], sizeof(_DownBuffer), SEGGER_RTT_MODE_NO_BLOCK_SKIP);
./ext/debug/segger/systemview/SEGGER_SYSVIEW.c:  _SYSVIEW_Globals.UpChannel = SEGGER_RTT_AllocUpBuffer  ("SysView", &_UpBuffer[0],   sizeof(_UpBuffer),   SEGGER_RTT_MODE_NO_BLOCK_SKIP);
./ext/debug/segger/systemview/SEGGER_SYSVIEW.c:  SEGGER_RTT_ConfigDownBuffer (_SYSVIEW_Globals.DownChannel, "SysView", &_DownBuffer[0], sizeof(_DownBuffer), SEGGER_RTT_MODE_NO_BLOCK_SKIP);
./ext/debug/segger/rtt/SEGGER_RTT_Conf.h:#define SEGGER_RTT_MODE_DEFAULT                   CONFIG_SEGGER_RTT_MODE
./ext/debug/segger/rtt/SEGGER_RTT.c:#ifndef   SEGGER_RTT_MODE_DEFAULT
./ext/debug/segger/rtt/SEGGER_RTT.c:  #define SEGGER_RTT_MODE_DEFAULT                         SEGGER_RTT_MODE_NO_BLOCK_SKIP
./ext/debug/segger/rtt/SEGGER_RTT.c:  p->aUp[0].Flags         = SEGGER_RTT_MODE_DEFAULT;
./ext/debug/segger/rtt/SEGGER_RTT.c:  p->aDown[0].Flags         = SEGGER_RTT_MODE_DEFAULT;
./ext/debug/segger/rtt/SEGGER_RTT.c:  case SEGGER_RTT_MODE_NO_BLOCK_SKIP:
./ext/debug/segger/rtt/SEGGER_RTT.c:  case SEGGER_RTT_MODE_NO_BLOCK_TRIM:
./ext/debug/segger/rtt/SEGGER_RTT.c:  case SEGGER_RTT_MODE_BLOCK_IF_FIFO_FULL:
./ext/debug/segger/rtt/SEGGER_RTT.c:  if (pRing->Flags == SEGGER_RTT_MODE_BLOCK_IF_FIFO_FULL) {
./ext/debug/segger/rtt/SEGGER_RTT.c:    if ((pRing->Flags & SEGGER_RTT_MODE_MASK) == SEGGER_RTT_MODE_BLOCK_IF_FIFO_FULL) {
./ext/debug/segger/rtt/SEGGER_RTT.c:    switch (pRing->Flags & SEGGER_RTT_MODE_MASK) {
./ext/debug/segger/rtt/SEGGER_RTT.c:    case SEGGER_RTT_MODE_NO_BLOCK_SKIP:
./ext/debug/segger/rtt/SEGGER_RTT.c:    case SEGGER_RTT_MODE_NO_BLOCK_TRIM:
./ext/debug/segger/rtt/SEGGER_RTT.c:    case SEGGER_RTT_MODE_BLOCK_IF_FIFO_FULL:
./ext/debug/segger/rtt/SEGGER_RTT.h:#define SEGGER_RTT_MODE_NO_BLOCK_SKIP         (0U)     // Skip. Do not block, output nothing. (Default)
./ext/debug/segger/rtt/SEGGER_RTT.h:#define SEGGER_RTT_MODE_NO_BLOCK_TRIM         (1U)     // Trim: Do not block, output as much as fits.
./ext/debug/segger/rtt/SEGGER_RTT.h:#define SEGGER_RTT_MODE_BLOCK_IF_FIFO_FULL    (2U)     // Block: Wait until there is space in the buffer.
./ext/debug/segger/rtt/SEGGER_RTT.h:#define SEGGER_RTT_MODE_MASK                  (3U)
```

I would have expected to see `CONFIG_SEGGER_RTT_MODE_` in the code.

Reply:
CONFIG_SEGGER_RTT_MODE

Reply:
@pizi-nordic I have just tried adding:

```
--- a/subsys/logging/Kconfig
+++ b/subsys/logging/Kconfig
@@ -298,6 +298,7 @@ config LOG_BACKEND_NATIVE_POSIX
 config LOG_BACKEND_RTT
        bool "Enable RTT backend"
        depends on RTT_CONSOLE
+       select SEGGER_RTT_MODE_NO_BLOCK_SKIP
        default y
        help
          When enabled backend is using RTT to output logs.
```

But the build system complains when I build `rm -Rf * && cmake -DBOARD=nrf52840_pca10056 ../samples/hello_world/` with the error:

```
warning: the choice symbol SEGGER_RTT_MODE_NO_BLOCK_SKIP (defined at ext/debug/segger/Kconfig:43) is selected by the following symbols, which has no effect: 
 - LOG_BACKEND_RTT (defined at subsys/logging/Kconfig:298)
```

I can see `SEGGER_RTT_MODE_NO_BLOCK_SKIP` is the default option in `ext/debug/segger/rtt/Kconfig` but even when I tried with a different value (for instance `SEGGER_RTT_MODE_BLOCK_IF_FIFO_FULL`), I have the same error. Any idea?